### PR TITLE
first pass at a dependency collector

### PR DESCRIPTION
**Pull request by jctanner**

### CODE REVIEW

File: migrate.py [hunk 1]
```diff
@@ -135,6 +136,50 @@
...
+        'unit': set(),
+        'integration': set(),
+    }
+
+    def __init__(self):
+        self.deps = {}
+
+    def add_import_dep(self, dep, plugin):
+        if dep not in self.deps:
+            self.deps[dep] = copy.deepcopy(self.NEWDEP)
```
Comment (webknjaz) [inline migrate.py]: Use a `defaultdict` instead.

File: migrate.py [hunk 2]
```diff
@@ -135,6 +136,50 @@
...
+
+    NEWDEP = {
+        'import': set(),
+        'doc': set(),
+        'unit': set(),
+        'integration': set(),
+    }
+
+    def __init__(self):
+        self.deps = {}
```
Comment (webknjaz) [inline migrate.py]: I believe you mean
```suggestion
        self.deps = defaultdict(defaultdict(set))
```

File: migrate.py [hunk 3]
```diff
@@ -135,6 +136,50 @@
 class UnmovablePathStr(str): ...
 
 
+class CollectionDependencyMap:
```
Comment (webknjaz) [inline migrate.py]: This doesn't have any hard deps on other things here, maybe it's worth putting into a separate module?

File: migrate.py [hunk 4]
```diff
@@ -135,6 +136,50 @@
...
+
+    def add_integration_dep(self, dep, plugin):
+        if plugin is None:
+            plugin = '!unknown'
+        if dep not in self.deps:
+            self.deps[dep] = copy.deepcopy(self.NEWDEP)
+        self.deps[dep]['integration'].add(plugin)
+
+    def to_safe_dict(self):
+        rdeps = {}
```
Comment (webknjaz) [inline migrate.py]:
```suggestion
        rdeps = defaultdict(dict)
```

File: migrate.py [hunk 5]
```diff
@@ -1851,19 +1880,21 @@ def rewrite_integration_tests(test_dirs, checkout_dir, collection_dir, namespace
                 if ext in BAD_EXT:
                     continue
                 elif ext in ('.py',):
-                    import_deps, docs_deps = rewrite_py(src, dest, collection, spec, namespace, args)
+                    import_deps, docs_deps = rewrite_py(src, dest, collection, spec, namespace, args, cdm=cdm)
 
                     for dep_ns, dep_coll in import_deps + docs_deps:
-                        integration_tests_add_to_deps((namespace, collection), (dep_ns, dep_coll))
+                        if cdm:
+                            cdm.add_integration_dep('.'.join([dep_ns, dep_coll]), dest)
```
Comment (webknjaz) [inline migrate.py]: just use an f-string
```suggestion
                            cdm.add_integration_dep(f'{dep_ns}.{dep_coll}', dest)
```

**Comment by jctanner** [comment]
This has been implemented elsewhere. awesome ...

**Pull request closed without merging**